### ISSUE TITLE
Saving SHAP plots programmatically in Python

### ISSUE THREAD

**Issue by MaxPowerWasTaken** [shap/shap#153]
First off, thanks a lot for such an awesome tool!

I think I might be missing something obvious, but I'm trying to save SHAP plots from Python, that I'm displaying with the shap plotting functions. I tried a couple ways:

```
import matplotlib.pyplot as plt
...
shap.summary_plot(shap_values, final_model_features)
plt.savefig('scratch.png')
```

and...

```
import matplotlib.pyplot as plt
...
fig = shap.summary_plot(shap_values, final_model_features)
plt.savefig('scratch.png')
```

but each just saves a blank image. Is there something obvious I'm missing to programmatically save these awesome plots from Python? Or should I just be re-generating them in matplotlib off the SHAP values matrix to do that? Thanks!

**Comment by slundberg** [comment]
By default `summary_plot` calls `plt.show()` to ensure the plot displays.
But if you pass `show=False` to `summary_plot` then it won't, and think
that might fix the blank figure issue you are having.

**Comment by MaxPowerWasTaken** [comment]
hey thanks! that did it.

**Comment by JoshKoh** [comment]
shap.summary_plot(shap_values, X, show=False)
This is not working for me. The plot still shows and plt.savefig only gives blank.

**Comment by slundberg** [comment]
@JoshKoh perhaps you have something like `%matplotlib inline` in your notebook? That might change things.

**Comment by hmanz** [comment]
Is there a way to store the figure into a variable instead of saving it to file?
Thanks

**Comment by slundberg** [comment]
@hmanz after running `shap.summary_plot(shap_values, X, show=False)` you can run `import matplotlib.pyplot as pl; f = pl.gcf()` to get the current figure in the variable f. What you do with it after that depends on matplotlib and not shap.

**Comment by hmanz** [comment]
@slundberg thank you for your great solution! I was able to use that solution for all SHAP plots except for the multiple-sample force-plots since there is no support for `matplotlib=True`. Is there a way to accomplish the same task for the multiple-sample force-plots?

More specifically, I am trying to store the plot as a variable so it can be handled by another class's method that will save it to file as a png or jpeg.

**Comment by slundberg** [comment]
Multiple-sample force-plots are Javascript/HTML. So in order to save those programmatically you will need to have some kind of browser-like renderer. Those exist but I can't comment on all the details to get them running.

**Comment by wabc1994** [comment]
Why that not work for me?

My code just like the follow, @slundberg 
the code just come from your official demo ,and i just want to save the demo image
```python
import xgboost
import shap
import matplotlib.pyplot as plt
# load JS visualization code to notebook
shap.initjs()
# train XGBoost model
X,y = shap.datasets.boston()
model = xgboost.train({"learning_rate": 0.01}, xgboost.DMatrix(X, label=y), 100)

# explain the model's predictions using SHAP values
# (same syntax works for LightGBM, CatBoost, and scikit-learn models)
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X)

# visualize the first prediction's explanation (use matplotlib=True to avoid Javascript)
# fig = plt.gcf()
# shap._*plot本质上就是调用 plt.show()， 所有我们的思路就是如何在下面的文件进行保存
shap.force_plot(explainer.expected_value, shap_values[0,:], X.iloc[0,:],show=False)
plt.savefig('scratch.png')
```

**Comment by dzieciou** [comment]
@wabc1994 I was able to save your plot by replacing the last two lines with:

    shap.force_plot(explainer.expected_value, shap_values[0,:], 
        X.iloc[0,:],show=False,matplotlib=True)
        .savefig('scratch.png')

Basically, it enables rendering with matplotlib instead of JavaScript and calls `savefig` on figure returned by `force_plot()`.

**Comment by paris007** [comment]
> ,show=False,matplotlib=True)
>     .savefig('scratch.png')

shap.force_plot(explainer.expected_value, shap_values, X,show=False,matplotlib=True).savefig('scratch.png')

I tried the code above and I get this error: 
Exception: matplotlib = True is not yet supported for force plots with multiple samples!

**Comment by sagydr** [comment]
`plt.savefig` worked for me. Thanks!
but.. can't I save it to **HTML** like I can after running force_plot?

`
shap.save_html('explainer.html', shap.force_plot(...))
`

**Comment by slundberg** [comment]
Note that when `matplotlib=True` is set then you can't save an HTML version because no HTML version exists (it is now a matplotlib figure, which is typically best exported as a PDF not HTML)

**Comment by charlatteD** [comment]
shap.force_plot(explainer.expected_value, shap_values[0,:], 
    X.iloc[0,:],show=False,matplotlib=True)
    .savefig('scratch.png') 

This works for me. But by specifying "matplotlib" = True, the resolution of the plot was downgraded, and a more serious issue is some parts of the original plot were cropped. Anyone had a similar issue?

**Comment by bradblanch** [comment]
@charlatteD this should solve your resolution and cropping issue:
shap.force_plot(explainer.expected_value, shap_values[0,:], X.iloc[0,:],show=False,matplotlib=True).savefig('scratch.png',format = "png",dpi = 150,bbox_inches = 'tight')

**Comment by Frederiklux** [comment]
I can save my force plots with:
shap.force_plot(explainer.expected_value[1], shap_values[1][0,:], top_11.iloc[0,:],show=False,matplotlib=True).savefig('test.png')

But when I include link="logit" as an argument, I get "posx and posy should be finite values" a bunch of times and a blank plot. Does anyone know how to solve this?

**Comment by lhang33** [comment]
> I can save my force plots with:
> shap.force_plot(explainer.expected_value[1], shap_values[1][0,:], top_11.iloc[0,:],show=False,matplotlib=True).savefig('test.png')
> 
> But when I include link="logit" as an argument, I get "posx and posy should be finite values" a bunch of times and a blank plot. Does anyone know how to solve this?

Same with you. Only that I found it occurs in version 0.36.0, but works well in 0.35.0. Hope it will be fixed sooner.

**Comment by giangnguyen2412** [comment]
When using plot_image, I can not save image because it does not return anything. I think plotting in shap is quite poor! Does someone have any experiences?

**Comment by harrytrinh2** [comment]
I got this error for force_plot `Exception: matplotlib = True is not yet supported for force plots with multiple samples!`

**Comment by Z-Shakeri** [comment]
The following workaround did the trick for me:

I **first** exported the interactive chart to HTML, **then** I saved the HTML file to a high-resolution pdf file.

Here is the code snippet I used:
```
f=shap.force_plot(explainer.expected_value, shap_values, X, show=False)
shap.save_html("index.htm", f)
```

**Comment by barrettab** [comment]
Has anyone been able to save either HTML or PNG file for a decision_plot?  I've been able to do this no problem for some of the other plots but haven't had any luck with this one specifically.

**Comment by singh0777** [comment]
@barrettab  Any luck with the decision plots? Thanks.

**Comment by singh0777** [comment]
> Has anyone been able to save either HTML or PNG file for a decision_plot? I've been able to do this no problem for some of the other plots but haven't had any luck with this one specifically.

Any luck with the decision plots? Thanks.

**Comment by AkankshaFarswan** [comment]
> > Has anyone been able to save either HTML or PNG file for a decision_plot? I've been able to do this no problem for some of the other plots but haven't had any luck with this one specifically.
> 
> Any luck with the decision plots? Thanks.

Yes, I am able to save decision plot as "png" using a simple argument "show=False" in the shap.decision_plot" command. Here is the code I used-

shap.decision_plot(explainer.expected_value, shap_values, show=False)
plt.savefig("decision.png")

**Comment by barrettab** [comment]
> > > Has anyone been able to save either HTML or PNG file for a decision_plot? I've been able to do this no problem for some of the other plots but haven't had any luck with this one specifically.
> > 
> > 
> > Any luck with the decision plots? Thanks.
> 
> Yes, I am able to save decision plot as "png" using a simple argument "show=False" in the shap.decision_plot" command. Here is the code I used-
> 
> shap.decision_plot(explainer.expected_value, shap_values, show=False)
> plt.savefig("decision.png")


@singh0777  This is the same code I used to make this work.

**Comment by varshakirani-cos** [comment]
If anyone is looking to display plot and also save the plot as an image, you can do this:
```
import matplotlib.pyplot as plt
...
fig=plt.gcf()
shap.summary_plot(shap_values, final_model_features)
fig.savefig('scratch.png')
```

**Comment by sorenwacker** [comment]
`savefig` works, but the labels are all cut off for long feature names. In that case use:

    savefig('test.svg', bbox_inches='tight')

**Comment by AlanConstantine** [comment]
> @hmanz after running `shap.summary_plot(shap_values, X, show=False)` you can run `import matplotlib.pyplot as pl; f = pl.gcf()` to get the current figure in the variable f. What you do with it after that depends on matplotlib and not shap.

doesn't work for me, my version is 0.40.0
```
import matplotlib.pyplot as pl
fig, axs = pl.subplots(1, 2, figsize=(22, 20))

shap.summary_plot(shap_values,
                  ......
                  show=False
                  )

axs[0] = pl.gcf()

axs[1].plot([1,2,3], [1,2,3])
```

![1](https://user-images.githubusercontent.com/24769943/142756781-a3e5b8cc-8f4f-4ef6-9654-64385a194665.png)

**Comment by Naviden** [comment]
Is there any solution for saving force_plot with multiple samples? Given that `matplotlib=True` is not supported ('0.40.0'), I'm doing `plt.savefig('scratch.png')`, which results in a blank image.

**Comment by sorenwacker** [comment]
Looks like is in fact working @AlanConstantine . Isn't that exactly what you would expect? If you want the summary plot in the first axis, then maybe call `pl.sca(axs[0])` before `shap.summary_plot`.

**Comment by JiaxiangBU** [comment]
I make a partial solution. Try to save all images outside and read them inside to merge. Here is a quick demo from https://stackoverflow.com/a/30228308/8625228 .

```
import sys
from PIL import Image

images = [Image.open(x) for x in ['Test1.jpg', 'Test2.jpg', 'Test3.jpg']]
widths, heights = zip(*(i.size for i in images))

total_width = sum(widths)
max_height = max(heights)

new_im = Image.new('RGB', (total_width, max_height))

x_offset = 0
for im in images:
  new_im.paste(im, (x_offset,0))
  x_offset += im.size[0]

new_im.save('test.jpg')
```

**Comment by AvivaSimchon** [comment]
> > I can save my force plots with:
> > shap.force_plot(explainer.expected_value[1], shap_values[1][0,:], top_11.iloc[0,:],show=False,matplotlib=True).savefig('test.png')
> > But when I include link="logit" as an argument, I get "posx and posy should be finite values" a bunch of times and a blank plot. Does anyone know how to solve this?
> 
> Same with you. Only that I found it occurs in version 0.36.0, but works well in 0.35.0. Hope it will be fixed sooner.

I have the same issue. only for XGBoost.
catBoost work well.
Does anyone know how to solve this?

**Comment by AvivaSimchon** [comment]
> > I can save my force plots with:
> > shap.force_plot(explainer.expected_value[1], shap_values[1][0,:], top_11.iloc[0,:],show=False,matplotlib=True).savefig('test.png')
> > But when I include link="logit" as an argument, I get "posx and posy should be finite values" a bunch of times and a blank plot. Does anyone know how to solve this?
> 
> Same with you. Only that I found it occurs in version 0.36.0, but works well in 0.35.0. Hope it will be fixed sooner.

I have the same issue

**Comment by azam-shakori** [comment]
> I can save my force plots with: shap.force_plot(explainer.expected_value[1], shap_values[1][0,:], top_11.iloc[0,:],show=False,matplotlib=True).savefig('test.png')
> 
> But when I include link="logit" as an argument, I get "posx and posy should be finite values" a bunch of times and a blank plot. Does anyone know how to solve this?

I have this problem, too

**Comment by mateusrn** [comment]
For those still trying to display and save a waterfall plot, this worked for me:

```
fig = plt.figure()
shap.plots.waterfall(shap_values, show=False)
plt.savefig(f'my_plot.png', bbox_inches='tight')
```
